tensorflow_probability/python/distributions/two_piece_normal.py

```diff
@@ -490,3 +393,301 @@ def _parameter_control_dependencies(self, is_init):
+
+  return tf.abs(normal_samples) * tf.where(
+      uniform_samples < tf.math.reciprocal(1. + skewness**2),
+      -tf.math.reciprocal(skewness),
+      skewness)
+
+
+def _two_piece_normal_sample_gradient(skewness, samples):
+  """Compute the gradients of Two-Piece Normal samples w.r.t. skewness.
+
+  This function computes the implicit reparameterization gradients [1]:
+
+  ```none
+  dz / dskewness = -(dF(z; skewness) / dskewness) / p(z; skewness)
+  ```
+
+  where `F(z; skewness)` and `p(z; skewness)` are the cdf and the pdf of the
+  Two-Piece Normal distribution with location zero, scale one, and skewness
+  `skewness`.
+
+  Args:
+    skewness: Floating-point tensor; the skewness(es) of the distribution(s).
+    samples: Floating-point tensor; the samples of the distribution(s).
+
+  Returns:
+    A tensor with shape broadcast according to the arguments.
+
+  Reference:
+    [1]: Michael Figurnov, Shakir Mohamed, and Andriy Mnih.
+         Implicit Reparameterization Gradients. In _Advances in Neural
+         Information Processing Systems_, 31, 2018.
+         https://arxiv.org/abs/1805.08498
+  """
+  one = tf.constant(1., dtype=skewness.dtype)
+  two = tf.constant(2., dtype=skewness.dtype)
+  pi = tf.constant(np.pi, dtype=skewness.dtype)
+  four = tf.constant(4., dtype=skewness.dtype)
+
+  left_piece = samples < 0.
+  z = samples * tf.where(left_piece, skewness, tf.math.reciprocal(skewness))
+
+  double_skewness = two * skewness
+  squared_skewness = tf.math.square(skewness)
+  scale = tf.math.reciprocal(one + squared_skewness)
+  squared_scale = tf.math.square(scale)
+
+  ndtr_term = four * skewness * _numpy_cast(
+      special_math.ndtr(z), skewness.dtype)
+  exp_term = scale * tf.math.sqrt(two / pi) * tf.math.exp(
+      -tf.math.square(z) / two)
+  samples_term = samples * exp_term
+
+  grad_left_piece = samples_term - squared_scale * ndtr_term
+  grad_right_piece0 = scale * (ndtr_term - double_skewness) - samples_term
+  grad_right_piece1 = squared_scale * (double_skewness + squared_skewness * (
+      ndtr_term - double_skewness))
+  grad_right_piece = grad_right_piece0 - grad_right_piece1
+
+  cdf_grad = tf.where(left_piece, grad_left_piece, grad_right_piece)
+  prob = skewness * exp_term
+
+  return -cdf_grad / prob
+
+
+def _two_piece_normal_sample_fwd(shape, skewness, seed):
+  """Compute output, aux (collaborates with _two_piece_normal_sample_bwd)."""
+  samples = _two_piece_normal_sample_no_gradient(shape, skewness, seed)
+  return samples, (skewness, samples)
+
+
+def _two_piece_normal_sample_bwd(_, aux, dy):
+  """The gradients of Two Piece Normal samples w.r.t. skewness."""
+  skewness, samples = aux
+  broadcast_skewness = tf.broadcast_to(skewness, ps.shape(samples))
+
+  grad = dy * _two_piece_normal_sample_gradient(broadcast_skewness, samples)
+  # Sum over the sample dimensions. Assume that they are always the first
+  # ones.
+  num_sample_dimensions = (tf.rank(broadcast_skewness) -
```

Comment:
In this function, you should use `ps.rank` and `ps.range`, as you're dealing with shapes and we want those to be statically propagated if possible.

Reply:
Done.